### PR TITLE
feat(query formatter) Add a Visitor to do Clickhouse specific formatting of Expressions

### PR DESCRIPTION
**Pull request by fpacifici**
This is the first step into the new Query AST formatting. This PR provides what we need to format Expressions. The next step will put them together into Query formatting.

In order to make formatting Clickhouse specific, this introduces a Visitor pattern (commonly used to evaluate asts) that is used to traverse the AST and format it node by node.
This formatter also manages escaping of each node type.

test plan:
- unit tests added

### CODE REVIEW

File: snuba/clickhouse/formatter.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+SPACE = " "
+TRUE = "true"
+FALSE = "false"
+QUOTE = "'"
+OPEN_BRACKET = "("
+CLOSED_BRACKET = ")"
+DOT = "."
+
+
+class ClickhouseExpressionFormatter(ExpressionVisitor[str]):
```
Comment (fpacifici) [inline snuba/clickhouse/formatter.py]: Hint: look at the unit test to see this working on corner cases.

File: snuba/clickhouse/formatter.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.parsing import ParsingContext
+from snuba.util import escape_alias, escape_col, escape_string
+
+# Tokens used when formatting. Defining them as costant
```
Comment (fpacifici) [inline snuba/clickhouse/formatter.py]: In the first implementation I was considering adding an abstraction for the Token. Looking at how simple the formatter actually is and knowing no further processing is supposed to happen on the formatted query, that seemed useless.

I think we could go back to this later if we wanted to support different ways of laying out the formatted query.

File: snuba/clickhouse/formatter.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
+
+    def __alias(self, formatted_exp: str, alias: Optional[str]) -> str:
+        if not alias:
+            return formatted_exp
+        elif self.__parsing_context.is_alias_present(alias):
+            ret = escape_alias(alias)
```
Comment (fpacifici) [inline snuba/clickhouse/formatter.py]: For now I relied on the existing escaping functions instead of providing a new implementation here. I don't think adding a second implementation (which would be simpler) gives us any advantage now.

File: snuba/query/expressions.py [hunk 4]
```diff
@@ -42,6 +44,46 @@ def __iter__(self) -> Iterator[Expression]:
...
+        """
+        Accepts a visitor class to traverse the tree. The only role of this method is to
+        call the right visit method on the visitor object. Requiring the implementation
+        to call the method with the right type forces us to keep the visitor interface
+        up to date every time we create a new subclass of Expression.
+        """
+        raise NotImplementedError
+
+
+class ExpressionVisitor(ABC, Generic[TVisited]):
```
Comment (fpacifici) [inline snuba/query/expressions.py]: For reference: https://en.wikipedia.org/wiki/Visitor_pattern

I looked into two possible designs for this:
1) The visitor pattern, which is commonly used to traverse and evaluate ASTs. The advantage is that we are sure the Expression class hierarchy cannot evolve without providing the needed changes to the formatter. 
2) Building a parallel hierarchy of Formatters in Clickhouse land. While that has some nice advantages (like being able to have a subclass of FunctionCall for Conditions only during formatting) it would be a lot more verbose and there would be no real way to ensure the two hierarchies evolve consistently.

Thus I went for the Visitor

File: snuba/query/expressions.py [hunk 5]
```diff
@@ -42,6 +44,46 @@ def __iter__(self) -> Iterator[Expression]:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def accept(self, visitor: ExpressionVisitor[TVisited]) -> TVisited:
```
Comment (fpacifici) [inline snuba/query/expressions.py]: Unfortunately with python and type hints we cannot overload methods thus `accept` has to be implemented in all subclasses.

File: snuba/clickhouse/formatter.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+CLOSED_BRACKET = ")"
+DOT = "."
+
+
+class ClickhouseExpressionFormatter(ExpressionVisitor[str]):
+    """
+    This Visitor implementation is able to format one expression in the Snuba
+    Query for Clickhouse.
+
+    The only state mainained is the PrasingContext, which allows us to resolve
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]:
```suggestion
    The only state maintained is the ParsingContext, which allows us to resolve
```

File: snuba/clickhouse/formatter.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    the visited expression, the return value is the formatted string.
+
+    This Formatter produces a properly escaped string. The result should never
+    be further escaped. This should be the only place where expression
+    escaping happens as it is done by each method that formats a specific
+    type of expression.
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: nit (extremely): Can this be an explicit `None` comparison instead of an `or` — just to be extra safe (paranoid)?

File: snuba/clickhouse/formatter.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        elif exp.value is True:
+            return TRUE
+        elif exp.value is False:
+            return FALSE
+        elif isinstance(exp.value, str):
+            return escape_string(exp.value)
+        elif isinstance(exp.value, (int, float)):
+            return str(exp.value)
+        else:
+            raise ValueError(f"Invalid literal type {type(exp.value)}")
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: It's possible that the possible types for `Literal.value` can change, this is probably better worded as:

```suggestion
            raise ValueError(f"Unexpected literal type: {type(exp.value)}")
```

File: snuba/clickhouse/formatter.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+# will make it easy (if/when needed) to make some changes
+# to the layout of the output.
+NULL = "NULL"
+COMMA = ","
+SPACE = " "
+TRUE = "true"
+FALSE = "false"
+QUOTE = "'"
+OPEN_BRACKET = "("
+CLOSED_BRACKET = ")"
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: These are parentheses, not brackets (which would be `[` and `]`.)

File: snuba/clickhouse/formatter.py (hunk 2 above)
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]:
```suggestion
# Tokens used when formatting. Defining them as constant
```

File: snuba/clickhouse/formatter.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+DOT = "."
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: I feel like this makes formatting more challenging understand at a glance, but if you think the flexibility is worth that disadvantage, feel free to leave as-is, I don't have a strong opinion one way or the other here.

File: snuba/clickhouse/formatter.py (hunk 7 above)
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: Kind of off-topic and/or out of scope, but we should remember to rename this class (`ParsingContext` when parsing and formatting are completely separate, since I don't think that this class will be used in parsing any longer.

File: snuba/clickhouse/formatter.py (hunk 3 above)
Comment (tkaemming) [reply snuba/clickhouse/formatter.py]: Makes sense, thanks for pointing that out, the comment below is helpful.

File: snuba/clickhouse/formatter.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    be further escaped. This should be the only place where expression
+    escaping happens as it is done by each method that formats a specific
+    type of expression.
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
+
+    def __alias(self, formatted_exp: str, alias: Optional[str]) -> str:
+        if not alias:
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: This has to handle both the empty string and `None`, right?

File: snuba/clickhouse/formatter.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.parsing import ParsingContext
+from snuba.util import escape_alias, escape_col, escape_string
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: Would it make sense to move some of these into this file (or at least this module) since I assume they are ClickHouse-specific (or at least only really intended for use with ClickHouse, as-is?)

File: snuba/clickhouse/formatter.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            return escape_string(exp.value)
+        elif isinstance(exp.value, (int, float)):
+            return str(exp.value)
+        else:
+            raise ValueError(f"Invalid literal type {type(exp.value)}")
+
+    def visitColumn(self, exp: Column) -> str:
+        ret = []
+        if exp.table_name:
+            ret.append(escape_col(exp.table_name) or "")
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: Is `escape_col` just named poorly? Should this be `escape_identifier` or something?

File: snuba/clickhouse/formatter.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+            ret.append(DOT)
+        ret.append(escape_col(exp.column_name) or "")
```
Comment (tkaemming) [inline snuba/clickhouse/formatter.py]: Why would `escape_col` return an empty value here?

File: tests/clickhouse/test_formatter.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+test_escaped = [
+    (
+        Column(None, "tags.values", "table.something"),
+        "table.something.tags.values",
+    ),  # Columns with dot are not escaped
+    (
+        Column(None, "tags[something]", "weird_!@#$%^^&*_table"),
+        "`weird_!@#$%^^&*_table`.`tags[something]`",
+    ),  # Somebody thought that table name was a good idea.
```
Comment (tkaemming) [inline tests/clickhouse/test_formatter.py]: lol

File: tests/clickhouse/test_formatter.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    (
+        Column("alias.cannot.have.dot", "columns.can", "table"),
+        "(table.columns.can AS `alias.cannot.have.dot`)",
+    ),  # Escaping is different between columns and aliases
+    (
+        FunctionCall(None, "f*&^%$#unction", [Column(None, "column", "table")]),
+        "`f*&^%$#unction`(table.column)",
+    ),  # Function names can be escaped. Hopefully it will never happen
```
Comment (tkaemming) [inline tests/clickhouse/test_formatter.py]: I didn't realize this was actually valid syntax, but looks like it is. Interesting.

File: tests/query/test_visitor.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    curried = CurriedFunctionCall("al7", f2, [f1])
+
+    visitor = DummyVisitor()
+    ret = curried.accept(visitor)
+
+    expected = [curried, f2, col2, literal2, f1, col1, literal1]
+
+    # Tests the state changes on the Visitor
+    assert visitor.get_visited_nodes() == expected
+    # Tests the retun value of the visitor
```
Comment (tkaemming) [inline tests/query/test_visitor.py]:
```suggestion
    # Tests the return value of the visitor
```

File: snuba/query/expressions.py (hunk 5 above)
Comment (tkaemming) [reply snuba/query/expressions.py]: I haven't played around with it much, but does [`typing.overload`](https://docs.python.org/3/library/typing.html#typing.overload) not work here?

**Review by tkaemming — COMMENTED** [review]
Mostly some nits and clarifying questions — overall this looks really good. I'd like to give another pass on `snuba/query/expressions.py` after I have a bit to think about the type structure.

File: snuba/clickhouse/formatter.py (hunk 9 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter.py]: mhmh, I knew my english teacher in primary school was lying about a lot of things.

File: snuba/clickhouse/formatter.py (hunk 10 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter.py]: Let me try a commit in which I keep only SPACE as a token. It will be trivial. To be fair I did not put too much thinking into this, I was thinking about a reversed tokenization. Maybe it is overkill.

File: snuba/clickhouse/formatter.py (hunk 11 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter.py]: True.

File: snuba/clickhouse/formatter.py (hunk 14 above)
Comment (fpacifici) [reply snuba/clickhouse/formatter.py]: It won't but the type is Optional because the first line of escape_expression is:
```
if not expr:
        return expr
```

File: snuba/query/expressions.py (hunk 5 above)
Comment (fpacifici) [reply snuba/query/expressions.py]: typing.overload does not do what I mean. typing.overload only allow you to define an alias for a function, in the end you have to collapse all the implementations of the overloaded signatures into one implementation method. Which here would defy the purpose. 
In Java/C++ I would be able to do something like:
```
class Visitor {
    void accept(Column e) {
            ...
    }
    void accept(FunctionCall f) {
           ...
    }
}

And the compiler does the appropriate dispatch, then it becomes useful because we can implement visit at the p[arent class level.

File: tests/clickhouse/test_formatter.py (hunk 16 above)
Comment (fpacifici) [reply tests/clickhouse/test_formatter.py]: I hope there is no clickhouse function that actually requires escaping. It would be ridiculous. So in practice this is unlikely to ever happen

**Comment by fpacifici** [comment]
Only architectural change is that I removed the token constants. The other changes are the nits from the code review

File: snuba/datasets/dataset.py [hunk 18]
```diff
@@ -1,13 +1,14 @@
 from typing import Optional, Mapping, NamedTuple, Sequence, Tuple, Union
 
+from snuba.clickhouse.escaping import escape_identifier
```
Comment (tkaemming) [inline snuba/datasets/dataset.py]: I don't know if this is a good thing to be excited about, but I do like how this makes the coupling here more explicit.

**Review by tkaemming — COMMENTED** [review]
I didn't really look in much depth, but I suspect some of the other `escape_` functions in `snuba/util.py` are also ClickHouse specific. Also, shouldn't the test(s) that had their import paths moved in `tests/test_util.py` be moved to a file that corresponds to their new path?

File: snuba/datasets/dataset.py (hunk 18 above)
Comment (fpacifici) [reply snuba/datasets/dataset.py]: I wanted to just move those method into the formatter, but then I would have util depending on the formatter, which would likely introduce a circular dependency as soon as the formatter needed something from util (which would be expected).

**Comment by fpacifici** [comment]
About the other escape_ functions in util. I left them there because they were not impacting this PR. I can move them as well.

**Review by tkaemming — APPROVED** [review]
Oops, didn't actually select the radio button on last review. 😬 

> About the other escape_ functions in util. I left them there because they were not impacting this PR. I can move them as well.

Your call, I think either doing that now or doing that as a follow up works.

File: snuba/datasets/dataset.py (hunk 18 above)
Comment (fpacifici) [reply snuba/datasets/dataset.py]: Let's not forge that the only dependency is in column_expr, thus it is going to disappear as soon the new query processing takes over.

Comment (tkaemming) [reply snuba/datasets/dataset.py]: Good point. I did mean this more generically, and as a positive, not as a criticism: I think that moving this (and other ClickHouse-specific functionality) to a module that literally says "`clickhouse`" in the module path makes it more clear where this coupling is so we can remove it in the future.
